File: debug_scripts/patch_final.py
import re
with open('C:\\Users\\MSI\\Desktop\\AI\\Odo\\index.html', 'r', encoding='utf-8') as f:
    code = f.read()

search = r"\} else if \(hasMatch\) \{\s*const codes = r\.ghnTripCode\.split\('\|'\)\.map\(c => c\.trim\(\)\)\.filter\(c => c\);\s*ghnHtml = <div style=\"display:flex; gap: 6px; align-items:center; flex-wrap:wrap;\">;\s*codes\.forEach\(\(c, idx\) => \{\s*ghnHtml \+= <a href=\"https://nhanh\.ghn\.vn/lastmile/trip-detail/\$\{c\}\" target=\"_blank\" style=\"color:var\(--success\);text-decoration:underline;\"><strong>\$\{c\}</strong></a>;\s*if \(idx < codes\.length - 1\) ghnHtml \+= <span style=\"color:#64748b;\">\|</span>;\s*\}\);\s*if \(r\.isManualMatch\) \{"

if re.search(search, code, re.DOTALL):
    print("ALREADY PATCHED!")
else:
    search2 = r"\} else if \(hasMatch\) \{.*?(?=if \(r\.isManualMatch\))"
    replace2 = r"""} else if (hasMatch) {
              const codes = String(r.ghnTripCode).split('|').map(c => c.trim()).filter(c => c);
              ghnHtml = <div style="display:flex; gap: 6px; align-items:center; flex-wrap:wrap;">;
              codes.forEach((c, idx) => {
                  ghnHtml += <a href="https://nhanh.ghn.vn/lastmile/trip-detail/" target="_blank" style="color:var(--success);text-decoration:underline;"><strong></strong></a>;
                  if (idx < codes.length - 1) ghnHtml += <span style="color:#64748b;">|</span>;
              });
              """
    code, n = re.subn(search2, lambda m: replace2, code, flags=re.DOTALL)
    if n > 0:
        with open('C:\\Users\\MSI\\Desktop\\AI\\Odo\\index.html', 'w', encoding='utf-8') as f:
            f.write(code)
        print("Patched successfully!")
    else:
        print("Still not found!")

File: debug_scripts/patch_final.js
const fs = require('fs');
let code = fs.readFileSync('C:/Users/MSI/Desktop/AI/Odo/index.html', 'utf8');

// The replacement logic:
const linesToReplace = [
    '<th style="width:60px;">Sửa</th>',
    '<th>Sửa</th>',
    '<th style="width:80px;">Sửa</th>',
    '<th>Chi tiết ngày chưa báo cáo</th>',
    '<th style="min-width:150px">Tuyến đường (Từ - Đến)</th>',
    '<td><a href="${sheetUrl}" target="_blank" style="color:var(--primary);text-decoration:underline;font-size:12px;white-space:nowrap;">Dòng ${e.sheetRow}</a></td>',
    '                            </div>\r\n                            </td>',
    '                            </div>\n                            </td>',
    '<td>${escapeHtml(String(e["Tuyến đường (Từ - Đến)"] || ""))}</td>'
];

let lines = code.split(/\r?\n/);

for (let i = 0; i < lines.length; i++) {
    const l = lines[i].trim();
    if (l === '<th style="width:60px;">Sửa</th>' || l === '<th>Sửa</th>' || l === '<th style="width:80px;">Sửa</th>' || l === '<th>Chi tiết ngày chưa báo cáo</th>' || l === '<th style="min-width:150px">Tuyến đường (Từ - Đến)</th>') {
        lines[i] = lines[i] + '\n<th style="width:40px; text-align:center;" title="Lịch sử thao tác">🕒</th>';
    }
    
    if (l === '<td><a href="${sheetUrl}" target="_blank" style="color:var(--primary);text-decoration:underline;font-size:12px;white-space:nowrap;">Dòng ${e.sheetRow}</a></td>') {
        lines[i] = lines[i] + '\n<td style="text-align:center;"><span style="opacity:0.2">🕒</span></td>';
    }
    
    if (l === '</div>' && lines[i+1].trim() === '</td>' && lines[i+2].trim() === '</tr>`;') {
        lines[i+1] = lines[i+1] + '\n<td style="text-align:center;"><span style="opacity:0.2">🕒</span></td>';
    }
    
    if (l === '<td>${escapeHtml(String(e["Tuyến đường (Từ - Đến)"] || ""))}</td>') {
        lines[i] = lines[i] + '\n<td style="text-align:center;"><span style="opacity:0.2">🕒</span></td>';
    }
}

fs.writeFileSync('C:/Users/MSI/Desktop/AI/Odo/index.html', lines.join('\n'));
console.log('Done4');
